collect/collect.go

```diff
@@ -762,6 +781,146 @@ func (i *InMemCollector) Stop() error {
 	return nil
 }
 
+// sentTrace is a struct that holds a trace and the record of the decision made.
+type sentTrace struct {
+	trace  *types.Trace
+	record cache.TraceSentRecord
+	reason string
+}
+
+// sendTracesInCache sends all traces in the cache to their final destination.
+// This is done on shutdown to ensure that all traces are sent before the collector
+// is stopped.
+func (i *InMemCollector) sendTracesInCache() {
+	wg := &sync.WaitGroup{}
+	if i.cache != nil {
+
+		// TODO: make this a config option
+		ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+		defer cancel()
+
+		traces := i.cache.GetAll()
+		sentTraceChan := make(chan sentTrace, len(traces)/3)
+		forwardTraceChan := make(chan *types.Trace, len(traces)/3)
+		expiredTraceChan := make(chan *types.Trace, len(traces)/3)
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
```

Comment:
I found this a bit confusing. Instead of calling `sendTracesOnShutdown` in a loop, I would prefer to see the for loop within that routine, which is a well-known pattern (for...select). Then the send function returns nothing and this goroutine is just a call to send and wg.Done.